File: php/src/Query.php
<?php

declare(strict_types=1);

namespace HopTop\Aim;

/** Query DSL parser. Mirrors Go `query.go`. */
final class Query
{
    private const KNOWN_TAG_KEYS = [
        'in', 'out', 'provider', 'family',
        'tool_call', 'reasoning', 'open_weights',
        'structured_output', 'temperature',
    ];

    public static function parse(string $q): Filter
    {
        $f = new Filter();
        $tokens = self::tokenise(trim($q));
        $free = [];

        foreach ($tokens as $tok) {
            if ($tok['quoted']) {
                $free[] = $tok['val'];
                continue;
            }
            $val = $tok['val'];
            $colonPos = strpos($val, ':');
            if ($colonPos === false) {
                $free[] = $val;
                continue;
            }
            $key = substr($val, 0, $colonPos);
            $rest = substr($val, $colonPos + 1);
            if ($key === '' || $rest === '') {
                throw new \InvalidArgumentException("aim: empty key or value around colon");
            }
            if (!in_array($key, self::KNOWN_TAG_KEYS, true)) {
                throw new \InvalidArgumentException("aim: unknown tag key \"$key\"");
            }
            self::applyTag($f, $key, $rest);
        }
        if ($free !== []) {
            $f->query = implode(' ', $free);
        }
        return $f;
    }

    /** @return list<array{val:string, quoted:bool}> */
    private static function tokenise(string $q): array
    {
        if ($q === '') return [];
        $tokens = [];
        $n = strlen($q);
        $i = 0;
        while ($i < $n) {
            $ch = $q[$i];
            if ($ch === ' ' || $ch === "\t") { $i++; continue; }
            if ($ch === '"') {
                $j = $i + 1;
                while ($j < $n && $q[$j] !== '"') $j++;
                if ($j >= $n) throw new \InvalidArgumentException("aim: unterminated quoted string in query");
                $tokens[] = ['val' => substr($q, $i + 1, $j - $i - 1), 'quoted' => true];
                $i = $j + 1;
                continue;
            }
            $start = $i;
            while ($i < $n && $q[$i] !== ' ' && $q[$i] !== "\t" && $q[$i] !== '"') $i++;
            $raw = substr($q, $start, $i - $start);
            if ($raw === ':') throw new \InvalidArgumentException("aim: bare colon in query");
            $tokens[] = ['val' => $raw, 'quoted' => false];
        }
        return $tokens;
    }

    private static function applyTag(Filter $f, string $key, string $val): void
    {
        match ($key) {
            'in'  => $f->input  = [...$f->input,  ...explode(',', $val)],
            'out' => $f->output = [...$f->output, ...explode(',', $val)],
            'provider' => $f->provider = $val,
            'family'   => $f->family   = $val,
            'tool_call'         => $f->toolCall         = self::parseBool($val),
            'reasoning'         => $f->reasoning         = self::parseBool($val),
            'open_weights'      => $f->openWeights       = self::parseBool($val),
            'structured_output' => $f->structuredOutput  = self::parseBool($val),
            'temperature'       => $f->temperature       = self::parseBool($val),
            default => throw new \InvalidArgumentException("aim: unknown tag key \"$key\""),
        };
    }

    private static function parseBool(string $s): bool
    {
        return match ($s) {
            'true' => true,
            'false' => false,
            default => throw new \InvalidArgumentException("aim: invalid bool value \"$s\": must be \"true\" or \"false\""),
        };
    }
}

File: rs/src/query.rs
//! Query string parser. Mirrors Go `query.go`.

use crate::types::Filter;

#[derive(Debug, thiserror::Error)]
#[error("aim: {0}")]
pub struct ParseError(pub String);

const KNOWN_TAG_KEYS: &[&str] = &[
    "in",
    "out",
    "provider",
    "family",
    "tool_call",
    "reasoning",
    "open_weights",
    "structured_output",
    "temperature",
];

struct Token {
    val: String,
    quoted: bool,
}

fn tokenise(q: &str) -> Result<Vec<Token>, ParseError> {
    let q = q.trim();
    if q.is_empty() {
        return Ok(vec![]);
    }
    let bytes = q.as_bytes();
    let n = bytes.len();
    let mut tokens = Vec::new();
    let mut i = 0;
    while i < n {
        match bytes[i] {
            b' ' | b'\t' => {
                i += 1;
            }
            b'"' => {
                let start = i + 1;
                let mut j = start;
                while j < n && bytes[j] != b'"' {
                    j += 1;
                }
                if j >= n {
                    return Err(ParseError("unterminated quoted string in query".into()));
                }
                tokens.push(Token {
                    val: String::from_utf8_lossy(&bytes[start..j]).into_owned(),
                    quoted: true,
                });
                i = j + 1;
            }
            _ => {
                let start = i;
                while i < n && !matches!(bytes[i], b' ' | b'\t' | b'"') {
                    i += 1;
                }
                let raw = String::from_utf8_lossy(&bytes[start..i]).into_owned();
                if raw == ":" {
                    return Err(ParseError("bare colon in query".into()));
                }
                tokens.push(Token {
                    val: raw,
                    quoted: false,
                });
            }
        }
    }
    Ok(tokens)
}

fn parse_bool(s: &str) -> Result<bool, ParseError> {
    match s {
        "true" => Ok(true),
        "false" => Ok(false),
        _ => Err(ParseError(format!(
            "invalid bool value \"{}\": must be \"true\" or \"false\"",
            s
        ))),
    }
}

fn apply_tag(f: &mut Filter, key: &str, val: &str) -> Result<(), ParseError> {
    match key {
        "in" => f.input.extend(val.split(',').map(String::from)),
        "out" => f.output.extend(val.split(',').map(String::from)),
        "provider" => f.provider = val.to_string(),
        "family" => f.family = val.to_string(),
        "tool_call" => f.tool_call = Some(parse_bool(val)?),
        "reasoning" => f.reasoning = Some(parse_bool(val)?),
        "open_weights" => f.open_weights = Some(parse_bool(val)?),
        "structured_output" => f.structured_output = Some(parse_bool(val)?),
        "temperature" => f.temperature = Some(parse_bool(val)?),
        _ => return Err(ParseError(format!("unknown tag key \"{}\"", key))),
    }
    Ok(())
}

pub fn parse_query(q: &str) -> Result<Filter, ParseError> {
    let mut f = Filter::default();
    let tokens = tokenise(q)?;
    let mut free = Vec::new();

    for tok in tokens {
        if tok.quoted {
            free.push(tok.val);
            continue;
        }
        if let Some((key, val)) = tok.val.split_once(':') {
            if key.is_empty() || val.is_empty() {
                return Err(ParseError("empty key or value around colon".into()));
            }
            if !KNOWN_TAG_KEYS.contains(&key) {
                return Err(ParseError(format!("unknown tag key \"{}\"", key)));
            }
            apply_tag(&mut f, key, val)?;
        } else {
            free.push(tok.val);
        }
    }

    if !free.is_empty() {
        f.query = free.join(" ");
    }
    Ok(f)
}
